radix-transaction-scenarios/src/scenarios/all_scenarios.rs

```diff
@@ -1,54 +1,422 @@
+            /* Callbacks */
+            on_transaction_execution: self.on_transaction_execution,
+            on_scenario_start: callback,
+        }
+    }
+
+    pub fn execute(mut self) -> Result<CompletedExecutionReceipt<D>, ScenarioExecutorError> {
+        // Bootstrapping if needed
+        if self.bootstrap {
+            Bootstrapper::new(
+                self.network_definition.clone(),
+                &mut self.database,
+                Vm::new(&self.scrypto_vm, self.native_vm.clone()),
+                false,
+            )
+            .bootstrap_test_default()
+            .ok_or(ScenarioExecutorError::BootstrapFailed)?;
+        };
+
+        // Running the scenarios.
+        let mut next_nonce = self.starting_nonce;
+        let mut scenario_builders = scenario_builders();
+        for requirement in self.scenarios_to_execute.clone() {
+            if let ScenarioRequirements::ProtocolUpdateUpTo(protocol_update) = requirement {
+                protocol_update
+                    .generate_state_updates(&self.database, &self.network_definition)
+                    .into_iter()
+                    .for_each(|state_updates| {
+                        self.database.commit(
+                            &state_updates.create_database_updates::<SpreadPrefixKeyMapper>(),
+                        )
+                    })
+            };
+
+            let scenario_builders = scenario_builders
+                .shift_remove(&requirement)
+                .unwrap_or_default();
+            for scenario_builder in scenario_builders {
+                let epoch = SystemDatabaseReader::new(&self.database)
+                    .read_object_field(
+                        CONSENSUS_MANAGER.as_node_id(),
+                        ModuleId::Main,
+                        ConsensusManagerField::State.field_index(),
+                    )
+                    .map_err(|_| ScenarioExecutorError::FailedToGetEpoch)?
+                    .as_typed::<VersionedConsensusManagerState>()
+                    .unwrap()
+                    .into_latest()
+                    .epoch;
+                let mut scenario = scenario_builder(ScenarioCore::new(
+                    self.network_definition.clone(),
+                    epoch,
+                    next_nonce,
+                ));
+                let metadata = scenario.metadata().clone();
+                (self.on_scenario_start)(&metadata);
+                let mut previous = None;
+                loop {
+                    let next = scenario
+                        .next(previous.as_ref())
+                        .map_err(|err| err.into_full(&scenario))
+                        .unwrap();
+                    match next {
+                        NextAction::Transaction(next) => {
+                            let receipt = self.execute_transaction(&next.raw_transaction)?;
+                            (self.on_transaction_execution)(
```

Comment:
I'd rather call this `on_transaction_executed`, because `on_transaction_execution` suggests to me it happens _before_ the transaction is executed.

Reply:
Sounds good, will rename this.